clang/lib/Driver/Driver.cpp

```diff
@@ -5203,761 +5225,806 @@ class OffloadingActionBuilder final {
+        }
+      };
 
-        if (A->getOption().matches(options::OPT_fsycl_targets_EQ)) {
-          // spir64 target is actually JIT compilation, so we defer selection of
-          // bfloat16 libraries to runtime. For AOT we need libraries, but skip
-          // for Nvidia.
-          needLibs = TC->getTriple().getSubArch() != llvm::Triple::NoSubArch &&
-                     !TC->getTriple().isNVPTX();
-          TargetBE = GetTripleIt(A->getValue(0));
-          if (TargetBE)
-            TargetOpt = A->getValue(0);
-          else
+      ActionList DeviceLibs;
+      ActionList DeviceLibObjects;
+      ActionList LinkObjects;
+      auto TT = TC->getTriple();
+      auto isNVPTX = TT.isNVPTX();
+      auto isAMDGCN = TT.isAMDGCN();
+      auto isSPIR = TT.isSPIR();
+      bool isSpirvAOT = TT.getSubArch() == llvm::Triple::SPIRSubArch_fpga ||
+                        TT.getSubArch() == llvm::Triple::SPIRSubArch_gen ||
+                        TT.getSubArch() == llvm::Triple::SPIRSubArch_x86_64;
+      const bool isSYCLNativeCPU =
+          TC->getAuxTriple() &&
+          driver::isSYCLNativeCPU(TT, *TC->getAuxTriple());
+      for (const auto &Input : LI) {
+        if (TT.getSubArch() == llvm::Triple::SPIRSubArch_fpga &&
+            types::isFPGA(Input->getType())) {
+          assert(BoundArch == nullptr &&
+                 "fpga triple bounded arch not nullptr");
+          // FPGA aoco does not go through the link, everything else does.
+          if (Input->getType() == types::TY_FPGA_AOCO) {
+            DeviceLibObjects.push_back(Input);
             continue;
-        } else if (A->getOption().matches(options::OPT_Xsycl_backend_EQ)) {
-          // Passing device args: -Xsycl-target-backend=<triple> <opt>
-          TargetBE = GetTripleIt(A->getValue(0));
-          if (TargetBE)
-            DeviceOpt = A->getValue(1);
+          }
+          // FPGA aocr/aocx does not go through the link and is passed
+          // directly to the backend compilation step (aocr) or wrapper (aocx)
+          Action *FPGAAOTAction;
+          if (Input->getType() == types::TY_FPGA_AOCR ||
+              Input->getType() == types::TY_FPGA_AOCR_EMU)
+            // Generate AOCX/AOCR
+            FPGAAOTAction =
+                C.MakeAction<BackendCompileJobAction>(Input, FPGAOutType);
+          else if (Input->getType() == types::TY_FPGA_AOCX)
+            FPGAAOTAction = Input;
           else
-            continue;
-        } else if (A->getOption().matches(options::OPT_Xsycl_backend)) {
-          // Passing device args: -Xsycl-target-backend <opt>
-          TargetBE = &SYCLTripleList.front();
-          DeviceOpt = A->getValue(0);
-        } else if (A->getOption().matches(options::OPT_Xs_separate)) {
-          // Passing device args: -Xs <opt>
-          DeviceOpt = A->getValue(0);
-        } else {
+            llvm_unreachable("Unexpected FPGA input type.");
+          auto *RenameAction = C.MakeAction<FileTableTformJobAction>(
```

Comment:
Can you please add some comment to help understand code between L5286 and L5292?